nebula_decoders/src/nebula_decoders_continental/decoders/continental_srr520_decoder.cpp

```diff
@@ -0,0 +1,1211 @@
+
+    uint16_t u_dist_x_std =
+      (static_cast<uint16_t>(data[11]) << 6) | (static_cast<uint16_t>(data[12] & 0xfc) >> 2);
+    assert(u_dist_x_std <= 16383);
+    object_msg.dist_x_std = DIST_STD_RESOLUTION * u_dist_x_std;
+
+    uint16_t u_dist_y_std = (static_cast<uint16_t>(data[12] & 0x03) << 12) |
+                            (static_cast<uint16_t>(data[13]) << 4) |
+                            (static_cast<uint16_t>(data[14] & 0xF0) >> 4);
+    assert(u_dist_y_std <= 16383);
+    object_msg.dist_y_std = DIST_STD_RESOLUTION * u_dist_y_std;
+
+    uint16_t u_v_abs_x_std = (static_cast<uint16_t>(data[14] & 0x0f) << 10) |
+                             (static_cast<uint16_t>(data[15]) << 2) |
+                             (static_cast<uint16_t>(data[15] & 0x03) >> 6);
+    assert(u_v_abs_x_std <= 16383);
+    object_msg.v_abs_x_std = V_ABS_STD_RESOLUTION * u_v_abs_x_std;
+
+    uint16_t u_v_abs_y_std =
+      (static_cast<uint16_t>(data[16] & 0x3f) << 8) | static_cast<uint16_t>(data[17]);
+    assert(u_v_abs_y_std <= 16383);
+    object_msg.v_abs_y_std = V_ABS_STD_RESOLUTION * u_v_abs_y_std;
+
+    uint16_t u_a_abs_x_std =
+      (static_cast<uint16_t>(data[18]) << 6) | (static_cast<uint16_t>(data[19] & 0xfc) >> 2);
+    assert(u_a_abs_x_std <= 16383);
+    object_msg.a_abs_x_std = A_ABS_STD_RESOLUTION * u_a_abs_x_std;
+
+    uint16_t u_a_abs_y_std = (static_cast<uint16_t>(data[19] & 0x03) << 12) |
+                             (static_cast<uint16_t>(data[20]) << 4) |
+                             (static_cast<uint16_t>(data[21] & 0xF0) >> 4);
+    assert(u_a_abs_y_std <= 16383);
+    object_msg.a_abs_y_std = A_ABS_STD_RESOLUTION * u_a_abs_y_std;
+
+    uint16_t u_box_length =
+      (static_cast<uint16_t>(data[21] & 0x0f) << 8) | static_cast<uint16_t>(data[22]);
+    assert(u_box_length <= 4095);
+    object_msg.box_length = OBJECT_BOX_RESOLUTION * u_box_length;
+
+    uint16_t u_box_width =
+      (static_cast<uint16_t>(data[23]) << 4) | (static_cast<uint16_t>(data[24] & 0xF0) >> 4);
+    assert(u_box_width <= 4095);
+    object_msg.box_width = OBJECT_BOX_RESOLUTION * u_box_width;
+
+    uint16_t u_orientation =
+      (static_cast<uint16_t>(data[24] & 0x0f) << 8) | static_cast<uint16_t>(data[25]);
+    assert(u_orientation <= 4094);
+    object_msg.orientation =
+      OBJECT_ORIENTATION_RESOLUTION * u_orientation + OBJECT_ORIENTATION_MIN_VALUE;
+
+    uint16_t u_rcs =
+      (static_cast<uint16_t>(data[26]) << 4) | (static_cast<uint16_t>(data[27] & 0xF0) >> 4);
+    assert(u_rcs <= 4094);
+    object_msg.rcs = OBJECT_RCS_RESOLUTION * u_rcs + OBJECT_RCS_MIN_VALUE;
+
+    uint8_t u_score = data[27] & 0x0f;
+    assert(u_score <= 15);
+    object_msg.score = OBJECT_SCORE_RESOLUTION * u_score;
+
+    object_msg.life_cycles = (static_cast<uint16_t>(data[28]) << 8) | data[29];
+    assert(object_msg.life_cycles <= 65535);
+
+    object_msg.box_valid = data[30] & 0x01;
+    object_msg.object_status = (data[30] & 0x06) >> 1;
```

Comment:
I wish C++ had nice un-aligned bitfield support.... then parsing code would be much cleaner.
But I agree that right now we can't really do anything nicer than this.